### ISSUE TITLE
decoupling components from context to enable async flows

### ISSUE THREAD

**Issue by yoshuawuyts** [segmentio/deku#51]
I've been playing with `deku` this weekend, and though I'm liking it a lot, I've ran into a few limitations. Internally it appears properties are  mutated in one of three ways:

1. By calling a function that `return`s a value (e.g. 
`component().prototype.render()`).
2. By calling functions bound to the scope (e.g. `this.setState()`).
3. By mutating function arguments (`state`, `props`).

This isn't much of a problem if you're using `deku` in a OOP way, but if you try to deviate from the model (for example by extending it with [event emitters](https://github.com/yoshuawuyts/deku-event)) this becomes problematic. There's been talk of stopping to [rely on `this` to mutate state (https://github.com/segmentio/deku/issues/37), but I'd like to suggest decoupling property mutation from the context even further to make all sorts of async flows possible.

An example for the `render` property:
```js
// OOP component
var button = deku.component()
button.prototype.render = function (props, state) {
  return deku.dom('button', state.text)
}

// decoupled component
var button = deku.component()
button.prototype.render = function (props, state, node) {
  node = deku.dom('button', state.text)  
}
```

The implementation of the above example would be completely backwards compatible, and plays well with [event emitters](https://github.com/yoshuawuyts/deku-event). Would you be willing to accept patches that normalize the mutation of component properties? Thanks!

**Comment by anthonyshort** [comment]
I'm not entirely sure how node is returned/used in your example. I want to get to the point when we're using a really functional API. Using events for `render` doesn't make much sense though, since there can only be one rendering of a component at any time. 

They make sense for other lifecycle events, and you could have plugins to start/stop timers and other things. But events really shouldn't mutate the state of the component itself. 

```js
button = component();

button.render(function(context){
  let {text} = context.props;
  return <div>{text}</div>;
});

button.afterMount(function(context){
  context = context.setState({ text: 'mounted' });
  return context;
});
```

There's no `this`, context is immutable, and every hook needs to return an updated context object. It's functional with no side-effects. This sort of API is preferable to an event API which relies on side-effects. To test components, you just pass a `context` object to a function and check the `context` object that is returned.

**Comment by anthonyshort** [comment]
If you're really after an event API then you're probably best to create a module that's just a wrapper around Deku.

**Comment by anthonyshort** [comment]
You could also allow multiple additions to the 

```
button
  .afterMount(fn)
  .afterMount(fn)
```

We could use events, but then the API is too generic. It needs to opinionated in some form for it to be useful and elegant.

We could allow them to be async but it could cause we're rendering issues:

```js
button.afterMount(function(context, next){
  context = context.setState({ text: 'mounted' });
  next(context);
});
```

**Comment by yoshuawuyts** [comment]
I like what you're proposing. Actually I already wrote [a wrapper](https://github.com/yoshuawuyts/deku-event/blob/master/index.js) for `deku.component()`, I prefer what you showed in your example. The syntax of the wrapper I wrote is something along the lines of:
```js
button.on('afterMount', (props, state) => {
  // do stuff
})
```
With the requirement to return a mutated `context` object, the wrapper I wrote wouldn't be possible to implement, but I'm 100% fine with that since it's indeed more predictable.

Using `.on('afterMount')` rather than `.afterMount()` might be a good way to syntactically differentiate between one-off functions such as `.render()` and events that can have multiple additions.

Closing this issue since it's mostly clear how this is going to be solved in the future. Thanks!
